baseplate/server/queue_consumer.py

```diff
@@ -0,0 +1,268 @@
+        while not self.stopped:
+            try:
+                # We set a timeout so we can periodically check if we should
+                # stop, this way we will actually return if we have recieved a
+                # `stop()` call rather than hanging in a `get` waiting for a
+                # new message.  If we did not do this, we would not be able to
+                # wait for all of our workers to finish before stopping the
+                # server.
+                message = self.work_queue.get(timeout=self._queue_timeout)
+            except queue.Empty:
+                pass
+            else:
+                # Ensure that if self.message_handler.handle throws a `queue.Empty`
+                # error, that bubbles up and is not treated as though `self.work_queue`
+                # is empty
+                self.message_handler.handle(message)
+        logger.debug("Consumer <%s> stopping.", self.id)
+
+
+class QueueConsumerServer:
+    """Server for running long-lived queue consumers."""
+
+    def __init__(
+        self, pump: PumpWorker, handlers: Sequence[QueueConsumer], healthcheck_server: StreamServer
+    ):
+        self.pump = pump
+        self.handlers = handlers
+        self.healthcheck_server = healthcheck_server
+        self.pump_thread = Thread(target=self.pump.run, daemon=True)
+        self.threads = [Thread(target=handler.run) for handler in self.handlers]
+        self.started = False
+        self.stopped = False
+        self._watch_period = 30
+
+    @classmethod
+    def new(
+        cls, max_concurrency: int, consumer_factory: QueueConsumerFactory, listener: socket.socket
+    ) -> "QueueConsumerServer":
+        """Build a new QueueConsumerServer."""
+        # We want to give some headroom on the queue so our handlers can grab
+        # a new message right after they finish so we keep an extra
+        # max_concurrency / 2 messages in the queue.
+        maxsize = max_concurrency + max_concurrency // 2
+        work_queue: queue.Queue = queue.Queue(maxsize=maxsize)
+        handlers = [
+            QueueConsumer(
+                work_queue=work_queue, message_handler=consumer_factory.build_message_handler()
+            )
+            for _ in range(max_concurrency)
+        ]
+        return cls(
+            pump=consumer_factory.build_pump_worker(work_queue),
+            handlers=handlers,
+            healthcheck_server=consumer_factory.build_health_checker(listener),
+        )
+
+    def _terminate(self) -> None:
+        """Send a SIGTERM signal to ourselves so baseplate can call `stop`."""
+        assert self.started
+        if not self.stopped:
+            os.kill(os.getpid(), signal.SIGTERM)
+
+    def _watch(self) -> None:
+        """Watches the threads to see if they have stopped prematurely.
+
+        Loops through the pump thread and handler threads periodically to check
+        if they are still alive and terminates the server if one is no longer
+        alive.
```

Comment:
sweeeeet

Reply:
FYI, I think there's probably a better way to do this that I'm gonna put up soon 😅 

```python
def watcher(fn):
    def _run_and_terminate(*a, **kw):
        try:
            return fn(*a, **kw)
        except Exception:
            self._terminate()

    return _run_and_terminate

self.pump_thread = Thread(target=watcher(self.pump.run), daemon=True)
self.threads = [Thread(target=watcher(handler.run)) for handler in self.handlers]
```